src/main/java/com/dumbdogdiner/stickyapi/common/book/chat/ClickEvent.java

```diff
@@ -0,0 +1,189 @@
+import lombok.Getter;
+import lombok.NonNull;
+
+/**
+ * An event that is fired when text associated with this event is clicked.
+ */
+public abstract class ClickEvent {
+    private ClickEvent() {} // seal this class
+
+    /**
+     * Convert the ClickEvent into JSON representation used in-game.
+     * @return {@link JsonElement}
+     */
+    public abstract @NonNull JsonElement toJson();
+
+    /**
+     * Create a copy of this ClickEvent.
+     * @return {@link ClickEvent}
+     */
+    public abstract @NonNull ClickEvent duplicate();
+
+    /**
+     * When clicked, displays a prompt to open a URL in the player's browser.
+     */
+    public final static class OpenUrl extends ClickEvent {
+        /** The URL to open. */
+        @Getter
+        private final @NonNull String url;
+
+        public OpenUrl(@NonNull String url) {
+            this.url = url;
+        }
+
+        @Override
+        public JsonElement toJson() {
+            JsonObject object = new JsonObject();
+            object.addProperty("action", "open_url");
+            object.addProperty("value", url);
+            return object;
+        }
+
+        @Override
+        public OpenUrl duplicate() {
+            return new OpenUrl(url);
+        }
+    }
+
+    /**
+     * When clicked, opens a file on the player's computer. This event can only be used by messages generated by the
+     * game, and will not function otherwise.
+     */
+    public final static class OpenFile extends ClickEvent {
+        /** The path to open. */
+        @Getter
+        private final @NonNull String path;
+
+        public OpenFile(@NonNull String path) {
+            this.path = path;
+        }
+
+        @Override
+        public JsonElement toJson() {
+            JsonObject object = new JsonObject();
+            object.addProperty("action", "open_file");
+            object.addProperty("value", path);
+            return object;
+        }
+
+        @Override
+        public OpenFile duplicate() {
+            return new OpenFile(path);
+        }
+    }
+
+    /**
+     * When clicked, executes the command as if the player said it in chat, either sending a message or executing a
+     * command. This can also be used to execute commands on signs, but only on the root component, with the slash
+     * omitted, and run by the server with @s set to the player.
+     */
+    public final static class RunCommand extends ClickEvent {
+        /** The command or message to run. */
+        @Getter
+        private final @NonNull String command;
+
+        public RunCommand(@NonNull String command) {
+            this.command = command;
+        }
+
+        @Override
+        public JsonElement toJson() {
+            JsonObject object = new JsonObject();
+            object.addProperty("action", "run_command");
+            object.addProperty("value", command);
+            return object;
+        }
+
+        @Override
+        public RunCommand duplicate() {
```

Comment:
I hate being a butt, but can we rename these to clone? ❤️

Reply:
Done